Data/Collections/Util/Array2DExtension.cs

```diff
@@ -0,0 +1,161 @@
+﻿using System;
+
+namespace Anvil.CSharp.Data
+{
+    //TODO: #104 - Optimize
+    /// <summary>
+    /// A set of helper extension methods for 2D arrays
+    /// </summary>
+    public static class Array2DExtension
+    {
+        /// <summary>
+        /// Gets the value at the specified index or returns the default value if out of range.
+        /// </summary>
+        /// <param name="array">The array to query.</param>
+        /// <param name="x">The first order index</param>
+        /// <param name="y">The second order index</param>
+        /// <typeparam name="T">The type of the array.</typeparam>
+        /// <returns>The value at the specified index or default.</returns>
+        public static T GetElementOrDefaultAt<T>(this T[,] array, int x, int y)
+        {
+            int limitX = array.GetLength(0);
+            int limitY = array.GetLength(1);
+
+            if (x >= limitX || y >= limitY || x < 0 || y < 0)
+            {
+                return default;
+            }
+
+            return array[x, y];
+        }
+
+        /// <summary>
+        /// Gets the index of the first element in the array that satisfies the specified <see cref="Predicate{T}"/>.
+        /// </summary>
+        /// <param name="array">The array to query.</param>
+        /// <param name="predicate">
+        /// The <see cref="Predicate{T}"/> to call for each element. If the predicate returns true the method returns
+        /// the current index.
+        /// </param>
+        /// <typeparam name="T">The type of the array.</typeparam>
+        /// <returns>The first index that satisfied the predicate.</returns>
+        public static (int,int) IndexOf<T>(this T[,] array, Predicate<T> predicate)
```

Comment:
I would suggest another name (maybe `FindIndexOf<T>`?) to indicate a conditional search - `IndexOf<T>` makes me think I'm just passing an element/value and asking for its index in the array.

Reply:
Good call. I'll follow the lead of `Linq` and `Array` and name the method `FindIndex`